File: sdks/go/qnsi/ai/ai.go
// Package ai is the QNSP AI Orchestrator client — model registry, AI
// workload submission with enclave attestation, inference, and bias /
// prompt-injection monitoring.
package ai

import (
	"context"
	"net/http"
	"time"

	"github.com/heossihq/qnsi-public/sdks/go/qnsi/internal/qnsicore"
)

const pathPrefix = "/proxy/ai/v1"

// Client is the QNSP AI Orchestrator client.
type Client struct {
	base *qnsicore.ServiceClient
}

// New constructs an AI client.
func New(act *qnsicore.Activator, httpClient *http.Client, timeout time.Duration) *Client {
	return &Client{base: qnsicore.NewServiceClient(act, httpClient, pathPrefix, timeout)}
}

// ── Model registry ──────────────────────────────────────────────────

// RegisterModelRequest is the JSON body for POST /ai/v1/models.
type RegisterModelRequest struct {
	Name         string         `json:"name"`
	Version      string         `json:"version"`
	Provider     string         `json:"provider"` // e.g. "openai", "anthropic", "local"
	Capabilities []string       `json:"capabilities,omitempty"`
	Metadata     map[string]any `json:"metadata,omitempty"`
}

func (c *Client) RegisterModel(ctx context.Context, req RegisterModelRequest, idempotencyKey string) (map[string]any, error) {
	return c.base.Do(ctx, http.MethodPost, "/models", req, nil, idempotencyKey)
}

func (c *Client) ListModels(ctx context.Context, query map[string]string) (map[string]any, error) {
	return c.base.Do(ctx, http.MethodGet, "/models", nil, query, "")
}

func (c *Client) GetModel(ctx context.Context, modelID string) (map[string]any, error) {
	return c.base.Do(ctx, http.MethodGet, "/models/"+modelID, nil, nil, "")
}

func (c *Client) UpdateModel(ctx context.Context, modelID string, body map[string]any, idempotencyKey string) (map[string]any, error) {
	return c.base.Do(ctx, http.MethodPatch, "/models/"+modelID, body, nil, idempotencyKey)
}

func (c *Client) ActivateModel(ctx context.Context, modelID string, idempotencyKey string) (map[string]any, error) {
	return c.base.Do(ctx, http.MethodPost, "/models/"+modelID+"/activate", nil, nil, idempotencyKey)
}

func (c *Client) DeployModel(ctx context.Context, body map[string]any, idempotencyKey string) (map[string]any, error) {
	return c.base.Do(ctx, http.MethodPost, "/models/deploy", body, nil, idempotencyKey)
}

// ── Workloads (long-running jobs in enclaves) ───────────────────────

// SubmitWorkloadRequest is the JSON body for POST /ai/v1/workloads.
type SubmitWorkloadRequest struct {
	ModelID    string         `json:"modelId"`
	Type       string         `json:"type"` // "training" | "fine-tune" | "inference-batch"
	InputRefs  []string       `json:"inputRefs,omitempty"`
	OutputBucket string       `json:"outputBucket,omitempty"`
	EnclaveType string        `json:"enclaveType,omitempty"` // e.g. "sgx", "sev", "nitro"
	Metadata   map[string]any `json:"metadata,omitempty"`
}

func (c *Client) SubmitWorkload(ctx context.Context, req SubmitWorkloadRequest, idempotencyKey string) (map[string]any, error) {
	return c.base.Do(ctx, http.MethodPost, "/workloads", req, nil, idempotencyKey)
}

func (c *Client) GetWorkload(ctx context.Context, workloadID string) (map[string]any, error) {
	return c.base.Do(ctx, http.MethodGet, "/workloads/"+workloadID, nil, nil, "")
}

func (c *Client) ListWorkloads(ctx context.Context, query map[string]string) (map[string]any, error) {
	return c.base.Do(ctx, http.MethodGet, "/workloads", nil, query, "")
}

func (c *Client) CancelWorkload(ctx context.Context, workloadID string, idempotencyKey string) (map[string]any, error) {
	return c.base.Do(ctx, http.MethodPost, "/workloads/"+workloadID+"/cancel", nil, nil, idempotencyKey)
}

// ── Inference ────────────────────────────────────────────────────────

// InferenceRequest is the JSON body for POST /ai/v1/inference.
type InferenceRequest struct {
	ModelID  string         `json:"modelId"`
	Input    map[string]any `json:"input"`
	Stream   bool           `json:"stream,omitempty"`
	Metadata map[string]any `json:"metadata,omitempty"`
}

func (c *Client) InvokeInference(ctx context.Context, req InferenceRequest, idempotencyKey string) (map[string]any, error) {
	return c.base.Do(ctx, http.MethodPost, "/inference", req, nil, idempotencyKey)
}

// ── Artifacts ────────────────────────────────────────────────────────

// RegisterArtifactRequest is the JSON body for POST /ai/v1/artifacts.
type RegisterArtifactRequest struct {
	Name      string         `json:"name"`
	Hash      string         `json:"hash"`
	StorageID string         `json:"storageId"`
	Type      string         `json:"type,omitempty"`
	Metadata  map[string]any `json:"metadata,omitempty"`
}

func (c *Client) RegisterArtifact(ctx context.Context, req RegisterArtifactRequest, idempotencyKey string) (map[string]any, error) {
	return c.base.Do(ctx, http.MethodPost, "/artifacts", req, nil, idempotencyKey)
}

// Do exposes the underlying service client for endpoints not yet wrapped.
func (c *Client) Do(ctx context.Context, method, path string, body any, query map[string]string, idempotencyKey string) (map[string]any, error) {
	return c.base.Do(ctx, method, path, body, query, idempotencyKey)
}

File: sdks/python/qnsi/src/qnsi/ai.py
"""QNSP AI Orchestrator - model registry, AI workload submission with
enclave attestation, inference, bias / prompt-injection monitoring."""

from __future__ import annotations

from typing import Any

from qnsi._service import _ServiceClient


class AIClient(_ServiceClient):
    """AI client. Wraps ``apps/ai-orchestrator`` (``/ai/v1``)."""

    PATH_PREFIX = "/ai/v1"

    # ── Model registry ──────────────────────────────────────────────

    def register_model(
        self,
        *,
        name: str,
        version: str,
        provider: str,
        capabilities: list[str] | None = None,
        metadata: dict[str, Any] | None = None,
        idempotency_key: str | None = None,
    ) -> dict[str, Any]:
        body: dict[str, Any] = {"name": name, "version": version, "provider": provider}
        if capabilities is not None:
            body["capabilities"] = capabilities
        if metadata is not None:
            body["metadata"] = metadata
        return self._request("POST", "/models", json=body, idempotency_key=idempotency_key)

    def list_models(self, **query: Any) -> dict[str, Any]:
        return self._request("GET", "/models", query=query or None)

    def get_model(self, model_id: str) -> dict[str, Any]:
        return self._request("GET", f"/models/{model_id}")

    def update_model(
        self,
        model_id: str,
        body: dict[str, Any],
        *,
        idempotency_key: str | None = None,
    ) -> dict[str, Any]:
        return self._request(
            "PATCH",
            f"/models/{model_id}",
            json=body,
            idempotency_key=idempotency_key,
        )

    def activate_model(
        self,
        model_id: str,
        *,
        idempotency_key: str | None = None,
    ) -> dict[str, Any]:
        return self._request(
            "POST",
            f"/models/{model_id}/activate",
            idempotency_key=idempotency_key,
        )

    def deploy_model(
        self,
        body: dict[str, Any],
        *,
        idempotency_key: str | None = None,
    ) -> dict[str, Any]:
        return self._request(
            "POST", "/models/deploy", json=body, idempotency_key=idempotency_key
        )

    # ── Workloads ───────────────────────────────────────────────────

    def submit_workload(
        self,
        *,
        model_id: str,
        type: str,  # noqa: A002 - matches API field name
        input_refs: list[str] | None = None,
        output_bucket: str | None = None,
        enclave_type: str | None = None,
        metadata: dict[str, Any] | None = None,
        idempotency_key: str | None = None,
    ) -> dict[str, Any]:
        body: dict[str, Any] = {"modelId": model_id, "type": type}
        if input_refs is not None:
            body["inputRefs"] = input_refs
        if output_bucket is not None:
            body["outputBucket"] = output_bucket
        if enclave_type is not None:
            body["enclaveType"] = enclave_type
        if metadata is not None:
            body["metadata"] = metadata
        return self._request(
            "POST", "/workloads", json=body, idempotency_key=idempotency_key
        )

    def get_workload(self, workload_id: str) -> dict[str, Any]:
        return self._request("GET", f"/workloads/{workload_id}")

    def list_workloads(self, **query: Any) -> dict[str, Any]:
        return self._request("GET", "/workloads", query=query or None)

    def cancel_workload(
        self,
        workload_id: str,
        *,
        idempotency_key: str | None = None,
    ) -> dict[str, Any]:
        return self._request(
            "POST",
            f"/workloads/{workload_id}/cancel",
            idempotency_key=idempotency_key,
        )

    # ── Inference ────────────────────────────────────────────────────

    def invoke_inference(
        self,
        *,
        model_id: str,
        input: dict[str, Any],  # noqa: A002 - matches API field name
        stream: bool = False,
        metadata: dict[str, Any] | None = None,
        idempotency_key: str | None = None,
    ) -> dict[str, Any]:
        body: dict[str, Any] = {"modelId": model_id, "input": input}
        if stream:
            body["stream"] = True
        if metadata is not None:
            body["metadata"] = metadata
        return self._request(
            "POST", "/inference", json=body, idempotency_key=idempotency_key
        )

    # ── Artifacts ────────────────────────────────────────────────────

    def register_artifact(
        self,
        *,
        name: str,
        hash: str,  # noqa: A002 - matches API field name
        storage_id: str,
        type: str | None = None,  # noqa: A002 - matches API field name
        metadata: dict[str, Any] | None = None,
        idempotency_key: str | None = None,
    ) -> dict[str, Any]:
        body: dict[str, Any] = {"name": name, "hash": hash, "storageId": storage_id}
        if type is not None:
            body["type"] = type
        if metadata is not None:
            body["metadata"] = metadata
        return self._request(
            "POST", "/artifacts", json=body, idempotency_key=idempotency_key
        )
